### ISSUE TITLE
Invalid warning when compiling with NTI in multiple threads

### ISSUE THREAD

**Issue by WearyMonkey** [google/closure-compiler#1925]
To reproduce:
```
public class Compile {
  private static final String CODE = "''.foo();";
  private static final String EXTERN = "" +
    "/**" +
    " * @constructor" +
    " */" +
    "function Function() {}" +
    "/**" +
    " * @constructor" +
    " */" +
    "function String() {}" +
    "String.prototype.foo = function() {};";

  public static void main(String[] args) throws InterruptedException {
    Thread thread1 = new Thread(Compile::compile);
    Thread thread2 = new Thread(Compile::compile);
    thread1.start();
    thread2.start();
    thread1.join();
    thread2.join();
  }

  private static void compile() {    
    CompilerOptions options = new CompilerOptions();
    options.setNewTypeInference(true);
    Compiler compiler = new Compiler(System.out);
    compiler.compile(SourceFile.fromCode("monkey", EXTERN), SourceFile.fromCode("banana", CODE), options);
  }
}
```

Expected Output: None

Actual Output:
```
banana:1: WARNING - Cannot call non-function type function(this:String):?
''.foo();
^

0 error(s), 1 warning(s), 100.0% typed
```

Root cause is the shared `JSType.commonTypes` [variable](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/newtypes/JSType.java#L80):
```
// NOTE(dimvar): This may cause problems when doing many compiles in the
// same Java process. But passing the JSTypes object around just to avoid
// making the field non-static is a huge readability pain.
private static JSTypes commonTypes = null;
```

As both threads will call `JSType.setCommonTypes` there is the potential for type instances created in one thread's compilation to leak into the other thread's compilation. For example, `JSType.autobox` returns types from the static `JSType.commonTypes` which if created on a different thread/compilation will be different instances than those in the local thread's `NewTypeInference.commonTypes`.

When `isSubtypeOf` [is called](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/NewTypeInference.java#L1828) on the the function type of  `''.foo` and `commonTypes.topFunction()` it returns false, as the two function types were created by different GlobalTypeInfo passes and therefore fail instance comparisons.

Possible to store `JSType.setCommonTypes` in a ThreadLocal? Or better yet remove the static variable entirely?

**Comment by ChadKillingsworth** [comment]
Related to #1822 

We don't really support multiple threads in the same JVM.

**Comment by dimvar** [comment]
It's not common to have multiple compiles in the same JVM, but we do support it, as long as there is a unique Compiler and unique CompilerOptions instance per compilation. I'll fix NTI for thread safety, but it's not high priority right now.

**Comment by supersteves** [comment]
I ran into this a while back when looking at a memory leak, #1822 as Chad refers to.

I found two cases of static (and potentially thread-unsafe) state:

* [`ObjectType.builtinObject`](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/newtypes/JSType.java#L80)
* [`JSType.commonTypes`](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/newtypes/ObjectType.java#L76)

For me, I'm just using reflection to clear memory prior to some unrelated memory leak detection logic. But it would be nice to be able to parallelise my several compilation jobs within the same JVM.

**Comment by dimvar** [comment]
Closed by cc1ea4c4a1f67dd637de504d8e96dd5f873e5667
